[1LP][RFR] Add test_infrastructure_hosts_refresh_multi to test_host.py

prichard77 opened the pull request:
##  Purpose or Intent
- __Updating tests__ for FA infra_hosts to cover refreshing multiple hosts. This was initially a manual test, but now it is automated.

{{ pytest: --use-provider complete --long-running cfme/tests/infrastructure/test_host.py::test_infrastructure_hosts_refresh_multi }}

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+):
+    """
+    Polarion:
+        assignee: pvala
```
Shouldn't you be the assignee for this test case?

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+               banner where "X" is the number of selected hosts. Properties for each host are
+               refreshed.
+    """
+    if len(provider.data.get('hosts', {})) < 2:
```
_Optional_ you can use `provider.hosts.all()` to get host objects. However, using data will be faster anyway.

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+    """
+    if len(provider.data.get('hosts', {})) < 2:
+        pytest.skip('not enough hosts to run test')
+    hosts_view = navigate_to(appliance.collections.hosts, "All")
```
_Optional_ I would just call this `view`

john-dupuy replied:
Also, I would navigate to the hosts through the provider object and not the appliance object, since this test will be parametrized against different providers.

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+    hosts_view = navigate_to(appliance.collections.hosts, "All")
+
+    assert hosts_view.is_displayed
+    [h.check() for h in hosts_view.entities.get_all()]
```
_Optional_ I'm not a fan of list comprehension without assignment, though more LOC, I think it is more readable to use a standard `for` loop.

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+
+    assert hosts_view.is_displayed
+    [h.check() for h in hosts_view.entities.get_all()]
+    num_checked_hosts = len(hosts_view.entities.get_all())
```
I would store this as a variable at the start of the test to reduce redundancy, it should always be the same as the number of hosts correct? (Unless a provider happened to have an inane no. of hosts on it)

john-dupuy commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,39 @@ def _finalize():
+    num_checked_hosts = len(hosts_view.entities.get_all())
+    hosts_view.toolbar.configuration.item_select('Refresh Relationships and Power States',
+                                                 handle_alert=True)
+    hosts_view.flash.assert_success_message(f'Refresh initiated for {num_checked_hosts} Hosts from'
```
_Optional_ formatting
```python,
    hosts_view.flash.assert_success_message(
        f'Refresh initiated for {num_checked_hosts} Hosts from the CFME Database'
    )
```

john-dupuy commented:
Also @prichard77 if you only want PRT to run this test you should specify that in the description block. I've updated your PR description for this PR.

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
     )
+
+
+@test_requirements.general_ui
```
I believe it should be `infra_hosts` FA. General UI is a different FA and has a different owner.

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+@test_requirements.general_ui
+def test_infrastructure_hosts_refresh_multi(
+    appliance, setup_provider, provider
+):
```
The test name and parameters fit into one line, please format in on one line.

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+        assignee: prichard
+        casecomponent: Infra
+        caseimportance: low
+        initialEstimate: 1/3h
```
Are you sure it take 20 minutes to refresh 2 hosts? Provider setup is not included in `initialEstimate`

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+        initialEstimate: 1/3h
+        testSteps:
+            1. Navigate to the Compute > Infrastructure > Hosts view.
+            2. Select at least 2 hosts (more if possible) by checking the box in upper left of
```
Optional: in the test I see that all hosts are refreshed but as it's clear what's going on, I leave it as optional

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+    if num_hosts < 2:
+        pytest.skip('not enough hosts to run test')
+    hosts_view = navigate_to(provider.collections.hosts, "All")
+    assert hosts_view.is_displayed
```
I don't think it's necessary here, if the navigation is not successful, it will fail on `is_displayed` anyway

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+        pytest.skip('not enough hosts to run test')
+    hosts_view = navigate_to(provider.collections.hosts, "All")
+    assert hosts_view.is_displayed
+    for h in hosts_view.entities.get_all():
```
I'm pretty sure we don't have that many hosts in any provider but what do you think about using `surf_pages=True`?

mshriver replied:
There is a method in the paginationpane widget for select all, I think we should use it here instead of iterating over like this.

Could be considered a separate test of pagination pane, could be parametrized, but I'm not sure that's worth it.

If we don't necessarily care how these items are selected, I'd use pagination select all, because its faster and a one-line call.

prichard77 replied:
This test is for refreshing multiple hosts, not all hosts. I'd like to leave the selection of the hosts similliar to the way it is, not using "all". I am going to take a closer look at how I determine the hosts (although similair code exists elsewhere). I could update the test to just use 2 hosts, but I figured we could test more if we had them and it would be a better test. Could add a refresh all test and use the suggested widget. I need to do some investigating as we are beginnig to go in circles with the PR.

lina-nikiforova commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+    hosts_view.toolbar.configuration.item_select('Refresh Relationships and Power States',
+                                                 handle_alert=True)
+    hosts_view.flash.assert_success_message(
+        f'Refresh initiated for {num_hosts} Hosts from the CFME Database'
```
I'm a bit concerned that refresh is made based on the number of hosts in the UI but `num_hosts` is determined based on the yaml file. Can you please explain?

lina-nikiforova replied:
@john-dupuy well, here my concern is that for some reason not all hosts are in yamls (e.g. somebody added some test host), that way this test will fail.

izapolsk commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,41 @@ def _finalize():
+    hosts_view.flash.assert_success_message(
+        f'Refresh initiated for {num_hosts} Hosts from the CFME Database'
+    )
+    assert wait_for(provider.is_refreshed, func_kwargs={"refresh_delta": 120}, num_sec=300,
```
you don't need assert here since it will raise Timedout exception

lina-nikiforova replied:
@izapolsk is it necessary to wrap? I agree with you on assert, but if we do just `wait_for(provider.is_refreshed...)` and it fails, the test fails as well.

prichard77 replied:
So we are not expecting each test to contain an assert? I am not arguing this point, just getting clarification. I am accustomed to each test having one, and only one assert (sometimes it makes sense to have more than one, but only in special cases). If the team is OK with tests passing with no asserts if they don't explicitly fail, I will comply.

prichard77 replied:
What is specifically wrong with the assert? Are we saying that the assert may cause/allow an incorrect test result, or do we just not like the explicit assert here. I'd definately want to fix anything that cases bad test results, but I don't see anything wrong with an explicit assert. It is what I am checking. If the wait_for times out, the test will fail. Otherwise the view is displayed and the test passes (exactly what I want). 
I have not tested the failure case (this test passes on the current builds), but the positive case seems to work. Do I need to cause the failure to make sure it would cause a fail result?

izapolsk replied:
@lina-nikiforova, there is a difference between Failed test and Error(Broken) test. I'm certain you understand the difference. From my point, yes, we need to wrap wait_for calls used to check test codition into try/expect and raise pytest.fail() in order to mark test as Failed not Broken.
btw, I don't say that we need to wrap all wait_for calls. I mean only wait_for calls in test which check some test condition.

@prichard77, let me explain what will happen there.
1. if condition is True in wait_for, it will pass True to your assert and test will be passed.

2. If condition in False in wait_for, wait_for will call that method until is reaches timeout value.
wait_for will raise TimedOutException in such case. So, your assert will never get False.
so, it's redundant. Btw, your test will be marked as Error(Broken) instead of Failed in such case.

So, what I think you need to do
```
try:
  wait_for(provider.is_refreshed, func_kwargs={"refresh_delta": 120}, num_sec=300, delay=30)
except TimedOutError:
   pytest.fail("Blahblah wasn't refreshed within given time")
```

lina-nikiforova replied:
@izapolsk thanks for the explanation! I agree with this point.

prichard77 replied:
I agree with the need to use try/except to better represent timeout as a failure and not error. I still question what is wrong with an assert? What about:
```
try:
  assert wait_for(provider.is_refreshed, func_kwargs={"refresh_delta": 120}, num_sec=300, delay=30)
except TimedOutError:
   pytest.fail("Blahblah wasn't refreshed within given time")

izapolsk replied:
@prichard77, there is nothing wrong with assert. it will just never be used in this particular case. because False will never be passed to it. so, it doesn't matter if it present or absent.

prichard77 commented:
Provider.is_refreashed is not doing what I wanted. I need to pretty much rewrite this test.

prichard77 commented:
The provider.is_refreshed is not doing anything useful for this test. It will give false positive. I have removed this from the test. I added a todo to add better checks. I spoke with Mike and he suggested looking for a "last refreshed" date for each host, storing it,  and comparing it after the refresh. But let's get this reviewed with only the check for the success banner. I may also add a check using the logValidator and checking evm logs.

dajoRH commented:
Lint report for commit c463f940c62c55d5d3cbe47ec5a7182e7da23296:

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 351:5: **E265** *block comment should start with '# '*
- :heavy_exclamation_mark: Line 351:98: **W292** *no newline at end of file*
- :heavy_exclamation_mark: Line 351:98: **W292** *no newline at end of file*
- :heavy_exclamation_mark: Line 351:98: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

izapolsk commented:
please make your test RFR after adding appropriate check.

mshriver commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -312,3 +312,40 @@ def _finalize():
+    """
+    hosts_view = navigate_to(provider.collections.hosts, "All")
+    num_hosts = len(hosts_view.entities.get_all())
+    if num_hosts < 2:
```
Just a thought, but not something to implement in this PR.

It would be great if ProviderFilter allowed for specifying a number of fields for required_fields. Right now it just supports looking for the presence of a field, and a matching value if supplied as a tuple.

For this PR, I think it would be better to use a fixture to get the number of hosts on the provider, skipping where there is only one. That way the skip happens in Setup, and not during Call phase.

The collection can actually be accessed directly via `provider.hosts`, an attribute is created by the BaseEntity class.

mshriver replied:
There's actually a subtle assertion going on here I guess, and that's the fact that the UI is displaying 2 or more hosts - its possible that the provider has more than two hosts in the REST collection used for `HostsCollection.all`, but there is  only one displayed on the UI.

prichard77 commented:
I am going to move the chack for number of hosts back up into a fixture. I'll check the actual provider data (not UI) before even starting the test on a provider. Then I'll check again (if this is a supported requirement) in the test, that we have the same number of hosts displayed. I'll integrate this somehow into the list I iterate over to check the hosts. It's been suggested that I use an "all" widget, but this test isn't specifically for all hosts (I may add one later) just for multiple hosts.

dajoRH commented:
I detected some fixture changes in commit a46829722c15be392f83bcad278ffd39fa80bb94

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 26583a1244dc3c910ba5b4abcd813ec51bba9b18

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
Lint report for commit 1b6a0cfd14942609476bf59650cb7a714835c943:

`cfme/tests/infrastructure/test_host.py`:
- :red_circle: Line 6:1: **F401** *'datetime' imported but unused*
- :red_circle: Line 353:5: **F841** *local variable 'current_refresh_date' is assigned to but never used*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit e60f2c9aa30e3a3cdd1d0a1e2be35be97472b0f1:

`cfme/tests/infrastructure/test_host.py`:
- :red_circle: Line 6:1: **F401** *'datetime' imported but unused*
- :red_circle: Line 353:5: **F841** *local variable 'current_refresh_date' is assigned to but never used*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
I detected some fixture changes in commit e60f2c9aa30e3a3cdd1d0a1e2be35be97472b0f1

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 5912a7ec53c7b77455bf25150d1e509063444175

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit 405789845dce0781e113719c3c042677cce671d6

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit ea7a074163ff467d8a5798a6e2bceda7acab69e4

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

dajoRH commented:
I detected some fixture changes in commit 1aa2afccab7a336615d75dff6beb5efcec1a6ff8

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

dajoRH commented:
I detected some fixture changes in commit c39a6034f140ea0226986d9f6813cd49dbb79c50

The *local* fixture **`setup_provider_min_hosts`** is used in the following files:
- **cfme/tests/infrastructure/test_host.py**
    - *test_infrastructure_hosts_refresh_multi*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

jawatts commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -331,3 +334,59 @@ def test_hosts_not_displayed_several_times(appliance, provider, setup_provider):
+    hosts_view = navigate_to(provider.collections.hosts, "All")
+    num_hosts = len(hosts_view.entities.get_all())
+    if num_hosts < 2:
+        pytest.skip('not enough hosts in appliance UI to run test')
```
This is duplicating the skip in the fixture above

jawatts replied:
Hmm looking at Mike's comment above, I see there is a second assertion going on here.

jawatts commented on cfme/common/provider.py:
```diff
@@ -639,7 +639,8 @@ def is_refreshed(self, refresh_timer=None, refresh_delta=600):
         if td > datetime.timedelta(0, refresh_delta):
-            self.refresh_provider_relationships()
+            if force_refresh:
+                self.refresh_provider_relationships()
```
This is a good enhancement

jawatts commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -331,3 +334,59 @@ def test_hosts_not_displayed_several_times(appliance, provider, setup_provider):
+    num_hosts = len(provider.data.get('hosts', {}))
+    if num_hosts < min_hosts:
+        pytest.skip(f'Not enough hosts({num_hosts}) to run test. Need at least {min_hosts}')
+    """Function-scoped fixture to set up a provider"""
```
**Optional** please use `#` is you want this comment in here

prichard77 replied:
I was thinking the same when I copied it. Though maybe you guys were doing it differently for some reason. I remember to use # for single line comments (as I usually do) going forward. Maybe I'll go back and fix this in the place I copied it from also in another PR.
